### ISSUE TITLE
A bug in modeling_tf_marian.py and modeling_tf_pegasus.py SinusoidalPositionalEmbedding _init_weight

### ISSUE THREAD

**Issue by JunnYu** [huggingface/transformers#11724]
## Information

should  create a new np.array and then store np.sin&np.cos result in it
```python
table = np.zeros_like(position_enc)
# index 0 is all zero
table[:, 0 : dim // 2] = np.sin(position_enc[:, 0::2])
table[:, dim // 2 :] = np.cos(position_enc[:, 1::2])
# convert to tensor
table = tf.convert_to_tensor(table)
tf.stop_gradient(table)
```
https://github.com/huggingface/transformers/blob/bd3b599c12cfcf5ef517c5ffe526afbdbaa92539/src/transformers/models/marian/modeling_tf_marian.py#L147-L157

https://github.com/huggingface/transformers/blob/8d43c71a1ca3ad322cc45008eb66a5611f1e017e/src/transformers/models/pegasus/modeling_tf_pegasus.py#L148-L158

**Comment by patrickvonplaten** [comment]
Hey @JunnYu,

Could you give a bit more context on what the issue is exactly? And how your solution solves the issue?

**Comment by JunnYu** [comment]
@patrickvonplaten 
After this line `position_enc[:, 0 : dim // 2] = np.sin(position_enc[:, 0::2]) ` and
` position_enc[:, 0 : dim // 2] ` value will be overridden in place.
If we compute  `np.cos(position_enc[:, 1::2])`  and the result is inconsistent with the expected result.
So we should  init a np.array `table = np.zeros_like(position_enc)`  to store the sinusoidalposition embeddings.

**Comment by patrickvonplaten** [comment]
True, good catch! Do you mind opening a PR to fix it? We will also have to check if this actually improves the model results

**Comment by JunnYu** [comment]
@patrickvonplaten  I have opened a PR  https://github.com/huggingface/transformers/pull/11897 to fix this.

**Comment by JunnYu** [comment]
@patrickvonplaten I have opened a PR #11897. Can you take the time to look at this PR？Thanks.

**Comment by patrickvonplaten** [comment]
Thanks for pinging me again & super sorry to have this on wait for so long!
